### ISSUE TITLE
design: how to delete PVCs/PVs without stopping DR protection

### ISSUE THREAD

**Issue by tjanssen3** [RamenDR/ramen#697]
This is a use case for CP4D. Opening an issue for tracking.

**Comment by pmuench1717** [comment]
If a user deletes a PVC that is protected by DR, Ramen holds off the deletion of the PVC via a finalizer.  Currently Ramen only removes the finalizer when DR is stopped (removing the VRG).  That doesn't work for Cloud Pak for Data (CPD) as PVCs can come and go during the lifecycle of CPD operation.  Disabling DR to remove a PVC is too heavyweight as CPD can have too many existing PVCs.  The data resync time could easily exceed the desired RPO of 30 minutes.  Plus the window of exposure to failure could also be unacceptable.

The challenge is how to communicate that the PVC is being deleted rather than the application being moved (failover or relocate) or terminated as these cases are already handled appropriately.  

Can we narrow the scope of the finalizer to just when it is actually needed?  I assume that CSI already handles the replication cleanup if a PV is deleted.

@ShyamsundarR Would you please add some more detail about the use and implementation of the finalizer.

**Comment by vdeenadh** [comment]
@ShyamsundarR @BenamarMk @pmuench1717 Here are my notes from our discussion on this issue.

### Acronyms
- ACM - Red Hat  #Advanced Cluster Management
- DRPC - Ramen DRPlacementControl CR on ACM hub cluster
- FDRO - Fusion Metro or Regional DR operator on DR cluster or user in some cases
- VRG - Ramen VolumeReplicationGroup CR on DR cluster
### Problem
While an application is DR protected, Ramen (as of Jan 2023) blocks deletion of PVCs (by adding finalizers to PVCs) until DR protection is stopped via deletion of DRPC/VRG. 
### Immediate cause 
A VRG in primary state does not allow garbage collection of any PVCs under its care until DRPC/FDRO stops DR protection by deleting the VRG. 
### Root cause
A VRG in primary state may experience the application going away in the following two different use cases, which requires the VRG to behave differently. 

- **Use case 1**: During a failover to a failover cluster, DRPC/FDRO may remove the application from the failed cluster while its VRG is in primary state.
  - **Required behavior**: In this case, the primary VRG (on the failed cluster) should not allow the PVC to be garbage collected because the underlying volume could be active on the peer cluster. 
- **Use case 2**: During normal application lifecycle, DRPC/FDRO may remove the application when it is no longer needed with the VRG still in primary state. 
  - **Required behavior**: In this case, the primary VRG (on the current cluster) should allow the PVC to be garbage collected.

Note the difference in the required behavior in the two use cases above. However, a primary VRG is unable to distinguish between the above two cases to behave differently and is hence, unable to handle the two cases correctly. As a work around, the current implementation gives preference to use case 1 over use case 2, where a primary VRG does not allow garbage collection of PVCs until the VRG itself is deleted. Because of this use case 2 is not currently supported. This enhancement is about supporting use case 2.

### Proposed Two-part Solution

1. DRPC/FDRO should remove the application from a failed cluster only after moving the VRG on the failed cluster to secondary state. This eliminates use case 1. With ACM based solution, in the failover scenario, the DRPC should do the following:
    - The PlacementRule should have both the failed cluster and failover cluster (instead of just the failover cluster) to prevent ACM from removing the application from the failed cluster while VRG on the failed cluster is in primary state.
    - When DRPC can talk to failed cluster, after it moves the VRG to secondary state (as it does today), it should then remove the failed cluster from the PlacementRule. This ensures that during failover scenario, ACM will remove the application from the failed cluster only after the VRG on the failed is in secondary state.
1. Having eliminated use case 1, a primary VRG has to only deal with use case 2 and hence, should allow garbage collection of PVCs even prior to deletion of the VRG. 

This proposal addresses the requirement documented in [Ramen issue #697](https://github.com/RamenDR/ramen/issues/697).

### Other use cases to consider

1. Will the system behave in a well-defined manner when PVC deletions and additions are made in GitOps during split brain scenario, where the VRG is in primary state on both the failed cluster and failover cluster?
1. What is Fusion documented procedure for application clean up in both of the above cases w.r.t VRG state?
1. How does this proposal affect Fusion’s use of ShioRamen?

### Summary of Proposed Behavior

**During failover and relocate operations**
In this case when it is not end-of-life of the application, clean up of application from the previous cluster should be serialized after its VRG has been moved to the secondary state. In other words, DRPC/FDRO/user should remove the application from the previous cluster only while the VRG is in the secondary state. Here the previous cluster refers to the relocation source cluster and to the failed-but-now-recovered cluster, both of which were previously housing the application (are no longer housing the application).

**During end-of-life of application**
 - DRPC/FDRO/user should remove the application from the current cluster only while the VRG is primary state. 
 - Warning note: Removing the application from the current cluster while the VRG is in secondary state (with no primary VRG on the peer cluster) will not result garbage collection of the PVCs (will result in leaking of storage volumes). This behavior is not changed by this proposal.

### Implementation Plan

- Given the urgency from Fusion perspective, first do the VRG changes under config map control.
- Do the DRPC changes at a good opportunity before CloudPak testing begins on ODF and make the VRG change the default behavior.

**Comment by ShyamsundarR** [comment]
Thanks @vdeenadh adding a couple of points when discussing the same with @hatfieldbrian as follows:
Mismatched hub and cluster Ramen version, with PVC as primary garbage collection enabled at the managed cluster
   - I.e Hub has older ramen operator and managed cluster has newer ramen operator
   - It maybe better for ramen config to start with this feature disabled, and let hub enable the feature (once updated to a supported version)
      - The hub may hence need to ensure the managed clusters are at a desired version that supports the feature, before enabling the same
   - Issues in regional maybe (if not ensured as above):
      - Failover app deletion racing with VRG as primary
         - This MAY detect VR as primary on both ends and hence not progress with the deletion. Current Primary detection though only relies on the VR `Completed` condition and needs some changes to ensure it is only primary on one end before deletion.
   - Issues in metro would be more dangerous (if not ensured as above):
      - The PVC and PV will be garbage collected on failover recovery, due to the above race

**Comment by hatfieldbrian** [comment]
# Design points
- When VRG controller is notified of a PVC deletion, reconcile VRG(s) that select it
- When VRG reconcile finds one of its selected PVCs has a non-zero deletion timestamp and a VRG finalizer, then unprotect it:
   - Delete its PV Kube object replica from VRG's S3 stores
      - A new s3 delete object routine is required because existing delete objects routine takes a prefix which may delete more PVs than the one to be deleted.  For example, deleting PV `b` also deletes PV `bb`.
   - Delete its VR if it has one
   - Change its reclaim policy back to delete if it VRG changed it
   - Remove its VRG annotations
   - Remove its VRG finalizer

## Assumptions
- A VRG's S3 store may not be removed
- PVC to VRG mapping:
   - A PVC may be selected by at most one VRG
   - A PVC's labels that result in a VRG selecting it are immutable
      - Orphaned PVCs could be garbage collected with a pointer to their former VRG with PR https://github.com/RamenDR/ramen/pull/627
   - A VRG's label selector is immutable

# Issues
## Issue 1 - Relocate
VRG relocate requires "to" and "from" clusters' VRGs' `Status.State` be `Secondary` so volume data can be synchronized.  For a VRG's state to become `Secondary` its PVCs must have a non-zero deletion timestamp.   Deleting a PVC would result in its unprotection, but PVCs must remain protected to monitor and report their data synchronization status.
## Solution approaches
### Approach 1 - Add another finalizer or condition to existing finalizer for relocate
How to know to add condition?  VRG Action?  PVC preservation must be in place before PVC is deleted.  Preservation condition should be removed once relocation completes.
### Approach 2 - Remove requirement that PVC be deleted
Why is this a requirement?  Volumes must not be updated by their PVCs for them to be synchronized.  For Regional DR, I presume that when a VR's `Status.State` is `Secondary` its volume is read-only.  Is there a generic way to make a PVC read-only that would work for Metro and Regional?

**Comment by hatfieldbrian** [comment]
@BenamarMk reminded me that the intent is to only unprotect deleted PVCs when VRG `Spec.State` is `Primary`.  This solves this issue above.

**Comment by hatfieldbrian** [comment]
Labeled high priority because the PR that addresses it is also.

**Comment by hatfieldbrian** [comment]
Fixed by #1116
